File: sdk/python/src/agentcc/_tokens.py
"""Token counting and cost estimation utilities."""

from __future__ import annotations

import json
from typing import Any

from agentcc._models_info import get_model_info


def encode(model: str, text: str) -> list[int]:
    """Encode text into token IDs using tiktoken.

    Args:
        model: Model name (used to select the tokenizer).
        text: Text to encode.

    Returns:
        List of token IDs.

    Raises:
        ImportError: If ``tiktoken`` is not installed.
    """
    try:
        import tiktoken
    except ImportError:
        raise ImportError(
            "tiktoken is required for encode(). "
            "Install it with: pip install tiktoken"
        ) from None

    try:
        enc = tiktoken.encoding_for_model(model)
    except KeyError:
        enc = tiktoken.get_encoding("cl100k_base")

    return enc.encode(text)


def decode(model: str, tokens: list[int]) -> str:
    """Decode token IDs back to text using tiktoken.

    Args:
        model: Model name (used to select the tokenizer).
        tokens: List of token IDs to decode.

    Returns:
        Decoded text string.

    Raises:
        ImportError: If ``tiktoken`` is not installed.
    """
    try:
        import tiktoken
    except ImportError:
        raise ImportError(
            "tiktoken is required for decode(). "
            "Install it with: pip install tiktoken"
        ) from None

    try:
        enc = tiktoken.encoding_for_model(model)
    except KeyError:
        enc = tiktoken.get_encoding("cl100k_base")

    return enc.decode(tokens)


def trim_messages(
    messages: list[dict[str, Any]],
    model: str,
    trim_ratio: float = 0.75,
    max_tokens: int | None = None,
) -> list[dict[str, Any]]:
    """Trim messages to fit within the model's context window.

    Keeps system messages and removes oldest user/assistant messages first
    until the total token count is within ``trim_ratio * max_tokens``.
    If *max_tokens* is provided, use that instead of model lookup.

    Args:
        messages: List of chat messages.
        model: Model name (used for tokenizer and context window lookup).
        trim_ratio: Fraction of the context window to use (default ``0.75``).
        max_tokens: Override context window size. If ``None``, looks up
            the model's max tokens from the model info database.

    Returns:
        Trimmed list of messages that fits within the token budget.

    Raises:
        ValueError: If the model is unknown and *max_tokens* is not provided.
    """
    if max_tokens is None:
        info = get_model_info(model)
        if info is None:
            raise ValueError(
                f"Unknown model {model!r}. "
                "Provide max_tokens explicitly or register the model first."
            )
        max_tokens = info.max_tokens

    budget = int(max_tokens * trim_ratio)

    # Check if messages already fit
    total = token_counter(model, messages=messages)
    if total <= budget:
        return list(messages)

    # Separate system messages from the rest
    system_msgs: list[dict[str, Any]] = []
    non_system_msgs: list[dict[str, Any]] = []
    for msg in messages:
        if msg.get("role") == "system":
            system_msgs.append(msg)
        else:
            non_system_msgs.append(msg)

    # Start with system messages (always kept)
    result = list(system_msgs)
    system_tokens = token_counter(model, messages=system_msgs) if system_msgs else 0
    remaining_budget = budget - system_tokens

    # Add non-system messages from newest to oldest until budget is exhausted
    kept: list[dict[str, Any]] = []
    running_tokens = 0
    for msg in reversed(non_system_msgs):
        msg_tokens = token_counter(model, messages=[msg])
        if running_tokens + msg_tokens <= remaining_budget:
            kept.append(msg)
            running_tokens += msg_tokens
        else:
            break

    # Reverse to restore chronological order and append after system messages
    kept.reverse()
    result.extend(kept)
    return result


def token_counter(
    model: str,
    text: str | None = None,
    messages: list[dict[str, Any]] | None = None,
) -> int:
    """Count tokens for text or messages.

    Uses ``tiktoken`` if installed, otherwise falls back to a character-based
    estimate (~4 chars per token for English text).

    Args:
        model: Model name (used to select the tokenizer).
        text: Raw text to count tokens for.
        messages: Chat messages list to count tokens for.

    Returns:
        Estimated token count.
    """
    if text is None and messages is None:
        return 0

    content = ""
    overhead = 0
    if text is not None:
        content = text
    elif messages is not None:
        # Serialize messages to estimate total tokens
        parts: list[str] = []
        for msg in messages:
            parts.append(msg.get("role", ""))
            msg_content = msg.get("content", "")
            if isinstance(msg_content, str):
                parts.append(msg_content)
            elif isinstance(msg_content, list):
                for part in msg_content:
                    if isinstance(part, dict) and part.get("type") == "text":
                        parts.append(part.get("text", ""))
            # Tool calls
            tool_calls = msg.get("tool_calls", [])
            if tool_calls:
                parts.append(json.dumps(tool_calls))
        content = "\n".join(parts)
        # Add overhead per message (~4 tokens per message for chat format)
        overhead = len(messages) * 4 + 2  # 2 for priming

    try:
        import tiktoken

        try:
            enc = tiktoken.encoding_for_model(model)
        except KeyError:
            enc = tiktoken.get_encoding("cl100k_base")

        count = len(enc.encode(content))
        if messages is not None:
            count += overhead
        return count
    except ImportError:
        # Fallback: ~4 characters per token for English
        count = len(content) // 4
        if messages is not None:
            count += overhead
        return max(count, 1)


def get_max_tokens(model: str) -> int | None:
    """Return the context window size for a model.

    Returns:
        Max input tokens, or ``None`` if model is unknown.
    """
    info = get_model_info(model)
    return info.max_tokens if info else None


def get_max_output_tokens(model: str) -> int | None:
    """Return the max output tokens for a model."""
    info = get_model_info(model)
    return info.max_output_tokens if info else None


def completion_cost(
    model: str,
    prompt_tokens: int = 0,
    completion_tokens: int = 0,
) -> float | None:
    """Estimate the cost of a completion.

    Args:
        model: Model name.
        prompt_tokens: Number of input tokens.
        completion_tokens: Number of output tokens.

    Returns:
        Estimated cost in USD, or ``None`` if model is unknown.
    """
    info = get_model_info(model)
    if info is None:
        return None
    return (prompt_tokens * info.input_cost_per_token) + (completion_tokens * info.output_cost_per_token)


# --- Context window fallbacks ---

CONTEXT_WINDOW_FALLBACKS: dict[str, str] = {
    "gpt-4": "gpt-4-turbo",
    "gpt-4-turbo": "gpt-4o",
    "gpt-3.5-turbo": "gpt-4o-mini",
    "claude-3-haiku-20240307": "claude-3-5-haiku-20241022",
    "claude-3-opus-20240229": "claude-sonnet-4-20250514",
}


def get_context_window_fallback(model: str) -> str | None:
    """Return a model with a larger context window, or ``None``."""
    return CONTEXT_WINDOW_FALLBACKS.get(model)


# --- Content policy fallbacks ---

CONTENT_POLICY_FALLBACKS: dict[str, str] = {
    "gpt-4o": "gpt-4-turbo",
    "gpt-4o-mini": "gpt-3.5-turbo",
}


def get_content_policy_fallback(model: str) -> str | None:
    """Return a less restrictive model for content policy errors, or ``None``."""
    return CONTENT_POLICY_FALLBACKS.get(model)


def is_prompt_caching_valid(model: str, messages: list[dict[str, Any]]) -> tuple[bool, str]:
    """Check if a prompt qualifies for provider-side caching.

    Inspects the model name and message structure to determine whether
    prompt caching can be leveraged.

    Args:
        model: The model name (e.g. ``'claude-3-5-sonnet-20241022'``, ``'gpt-4o'``).
        messages: Chat messages list.

    Returns:
        Tuple of (is_eligible, reason_string).
    """
    # Check for Anthropic models with cache_control
    if "claude" in model.lower():
        for msg in messages:
            if "cache_control" in msg:
                return (True, "Anthropic cache_control detected")
            # Also check content parts for cache_control
            content = msg.get("content")
            if isinstance(content, list):
                for part in content:
                    if isinstance(part, dict) and "cache_control" in part:
                        return (True, "Anthropic cache_control detected")

    # Check for OpenAI models with long system prompts
    if any(prefix in model.lower() for prefix in ("gpt-", "o1", "o3")):
        for msg in messages:
            if msg.get("role") == "system":
                system_content = msg.get("content", "")
                if isinstance(system_content, str):
                    system_tokens = token_counter(model, text=system_content)
                    if system_tokens >= 1024:
                        return (True, "System prompt eligible for OpenAI automatic caching")

    return (False, "No caching indicators found")


def completion_cost_from_response(response: Any) -> float | None:
    """Estimate cost from a ChatCompletion response object.

    First checks ``response.agentcc.cost`` (actual cost from gateway),
    falling back to local estimation if not available.
    """
    # Try gateway-reported cost first
    agentcc = getattr(response, "agentcc", None)
    if agentcc and getattr(agentcc, "cost", None) is not None:
        return agentcc.cost  # type: ignore[no-any-return]

    # Fall back to estimation
    usage = getattr(response, "usage", None)
    model = getattr(response, "model", None)
    if usage and model:
        return completion_cost(
            model=model,
            prompt_tokens=getattr(usage, "prompt_tokens", 0),
            completion_tokens=getattr(usage, "completion_tokens", 0),
        )
    return None

File: sdk/typescript/src/tokens.ts
/**
 * Token counting, cost estimation, and prompt utilities.
 *
 * Token counting uses a character-based estimate (~4 chars per token for
 * English text).  This is intentionally dependency-free — if exact counts
 * are needed, pass the text through tiktoken separately.
 *
 * @module
 */

import { getModelInfo } from "./models-info.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ChatMessage {
  role?: string;
  content?: string | ContentPart[] | null;
  tool_calls?: unknown[];
  [key: string]: unknown;
}

interface ContentPart {
  type?: string;
  text?: string;
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Token counting
// ---------------------------------------------------------------------------

/**
 * Estimate the number of tokens for text or messages.
 *
 * Uses a character-based heuristic (~4 English characters per token).
 * Add `tiktoken` or `gpt-tokenizer` in your project for exact counts.
 */
export function tokenCounter(
  model: string,
  opts: { text?: string; messages?: ChatMessage[] } = {},
): number {
  const { text, messages } = opts;
  if (text == null && messages == null) return 0;

  let content = "";
  let overhead = 0;

  if (text != null) {
    content = text;
  } else if (messages != null) {
    const parts: string[] = [];
    for (const msg of messages) {
      parts.push(msg.role ?? "");
      const c = msg.content;
      if (typeof c === "string") {
        parts.push(c);
      } else if (Array.isArray(c)) {
        for (const part of c) {
          if (part && typeof part === "object" && part.type === "text") {
            parts.push(part.text ?? "");
          }
        }
      }
      if (msg.tool_calls) {
        parts.push(JSON.stringify(msg.tool_calls));
      }
    }
    content = parts.join("\n");
    overhead = messages.length * 4 + 2; // ~4 tokens per message + 2 priming
  }

  // Character-based estimate: ~4 chars per token
  const count = Math.max(Math.ceil(content.length / 4), text != null || messages != null ? 1 : 0);
  return count + overhead;
}

// ---------------------------------------------------------------------------
// Context window queries
// ---------------------------------------------------------------------------

/** Return the context window size for a model, or `null` if unknown. */
export function getMaxTokens(model: string): number | null {
  return getModelInfo(model)?.maxTokens ?? null;
}

/** Return the max output tokens for a model, or `null` if unknown. */
export function getMaxOutputTokens(model: string): number | null {
  return getModelInfo(model)?.maxOutputTokens ?? null;
}

// ---------------------------------------------------------------------------
// Cost estimation
// ---------------------------------------------------------------------------

/**
 * Estimate the cost of a completion in USD.
 *
 * Returns `null` if the model is unknown.
 */
export function completionCost(
  model: string,
  promptTokens: number = 0,
  completionTokens: number = 0,
): number | null {
  const info = getModelInfo(model);
  if (!info) return null;
  return (
    promptTokens * info.inputCostPerToken +
    completionTokens * info.outputCostPerToken
  );
}

/**
 * Estimate cost from a ChatCompletion-like response object.
 *
 * Checks `response.agentcc.cost` (gateway-reported) first, then falls back
 * to local estimation from usage data.
 */
export function completionCostFromResponse(response: {
  model?: string;
  usage?: { prompt_tokens?: number; completion_tokens?: number };
  agentcc?: { cost?: number | null };
}): number | null {
  // Gateway-reported cost
  if (response.agentcc?.cost != null) {
    return response.agentcc.cost;
  }

  // Local estimation
  if (response.usage && response.model) {
    return completionCost(
      response.model,
      response.usage.prompt_tokens ?? 0,
      response.usage.completion_tokens ?? 0,
    );
  }
  return null;
}

// ---------------------------------------------------------------------------
// Message trimming
// ---------------------------------------------------------------------------

/**
 * Trim messages to fit within a model's context window.
 *
 * Always preserves system messages.  Removes oldest non-system messages
 * first until the total token estimate is within `trimRatio * maxTokens`.
 *
 * @throws {Error} If the model is unknown and `maxTokens` is not provided.
 */
export function trimMessages(
  messages: ChatMessage[],
  model: string,
  opts: { trimRatio?: number; maxTokens?: number } = {},
): ChatMessage[] {
  const { trimRatio = 0.75 } = opts;
  let { maxTokens } = opts;

  if (maxTokens == null) {
    const info = getModelInfo(model);
    if (!info) {
      throw new Error(
        `Unknown model "${model}". ` +
          "Provide maxTokens explicitly or register the model first.",
      );
    }
    maxTokens = info.maxTokens;
  }

  const budget = Math.floor(maxTokens * trimRatio);

  // Check if messages already fit
  const total = tokenCounter(model, { messages });
  if (total <= budget) return [...messages];

  // Separate system from non-system messages
  const systemMsgs: ChatMessage[] = [];
  const nonSystemMsgs: ChatMessage[] = [];
  for (const msg of messages) {
    if (msg.role === "system") {
      systemMsgs.push(msg);
    } else {
      nonSystemMsgs.push(msg);
    }
  }

  const systemTokens = systemMsgs.length
    ? tokenCounter(model, { messages: systemMsgs })
    : 0;
  const remainingBudget = budget - systemTokens;

  // Add non-system messages from newest to oldest until budget exhausted
  const kept: ChatMessage[] = [];
  let runningTokens = 0;
  for (let i = nonSystemMsgs.length - 1; i >= 0; i--) {
    const msgTokens = tokenCounter(model, { messages: [nonSystemMsgs[i]] });
    if (runningTokens + msgTokens <= remainingBudget) {
      kept.push(nonSystemMsgs[i]);
      runningTokens += msgTokens;
    } else {
      break;
    }
  }
  kept.reverse();

  return [...systemMsgs, ...kept];
}

// ---------------------------------------------------------------------------
// Context window & content policy fallbacks
// ---------------------------------------------------------------------------

const CONTEXT_WINDOW_FALLBACKS: Record<string, string> = {
  "gpt-4": "gpt-4-turbo",
  "gpt-4-turbo": "gpt-4o",
  "gpt-3.5-turbo": "gpt-4o-mini",
  "claude-3-haiku-20240307": "claude-3-5-haiku-20241022",
  "claude-3-opus-20240229": "claude-sonnet-4-20250514",
};

/** Return a model with a larger context window, or `null`. */
export function getContextWindowFallback(model: string): string | null {
  return CONTEXT_WINDOW_FALLBACKS[model] ?? null;
}

const CONTENT_POLICY_FALLBACKS: Record<string, string> = {
  "gpt-4o": "gpt-4-turbo",
  "gpt-4o-mini": "gpt-3.5-turbo",
};

/** Return a less restrictive model for content policy errors, or `null`. */
export function getContentPolicyFallback(model: string): string | null {
  return CONTENT_POLICY_FALLBACKS[model] ?? null;
}

// ---------------------------------------------------------------------------
// Prompt caching detection
// ---------------------------------------------------------------------------

/**
 * Check if a prompt qualifies for provider-side caching.
 *
 * Returns `[isEligible, reason]`.
 */
export function isPromptCachingValid(
  model: string,
  messages: ChatMessage[],
): [boolean, string] {
  // Anthropic models with cache_control
  if (model.toLowerCase().includes("claude")) {
    for (const msg of messages) {
      if ("cache_control" in msg) {
        return [true, "Anthropic cache_control detected"];
      }
      const content = msg.content;
      if (Array.isArray(content)) {
        for (const part of content) {
          if (part && typeof part === "object" && "cache_control" in part) {
            return [true, "Anthropic cache_control detected"];
          }
        }
      }
    }
  }

  // OpenAI models with long system prompts
  const lower = model.toLowerCase();
  if (["gpt-", "o1", "o3"].some((p) => lower.includes(p))) {
    for (const msg of messages) {
      if (msg.role === "system" && typeof msg.content === "string") {
        const systemTokens = tokenCounter(model, { text: msg.content });
        if (systemTokens >= 1024) {
          return [true, "System prompt eligible for OpenAI automatic caching"];
        }
      }
    }
  }

  return [false, "No caching indicators found"];
}
